app/upgrades.go

```diff
@@ -53,119 +40,6 @@ type appUpgrade struct {
-
-			return vm, nil
-		},
-		Added: []string{icqtypes.ModuleName, oracletypes.ModuleName, ibchookstypes.StoreKey, hold.ModuleName, exchange.ModuleName},
-	},
-	"saffron-rc2": { // upgrade for v1.17.0-rc2
-		Handler: func(ctx sdk.Context, app *App, vm module.VersionMap) (module.VersionMap, error) {
-			var err error
-			vm, err = runModuleMigrations(ctx, app, vm)
-			if err != nil {
-				return nil, err
-			}
-
-			updateIbcMarkerDenomMetadata(ctx, app)
-
-			return vm, nil
-		},
-	},
-	"saffron-rc3": { // upgrade for v1.17.0-rc3
-		Handler: func(ctx sdk.Context, app *App, vm module.VersionMap) (module.VersionMap, error) {
-			var err error
-			vm, err = runModuleMigrations(ctx, app, vm)
-			if err != nil {
-				return nil, err
-			}
-
-			updateIbcMarkerDenomMetadata(ctx, app)
-
-			return vm, nil
-		},
-	},
-	"saffron": { // upgrade for v1.17.0,
-		Handler: func(ctx sdk.Context, app *App, vm module.VersionMap) (module.VersionMap, error) {
-			var err error
-			vm, err = runModuleMigrations(ctx, app, vm)
-			if err != nil {
-				return nil, err
-			}
-
-			// set ibchoooks defaults (no allowed async contracts)
-			app.IBCHooksKeeper.SetParams(ctx, ibchookstypes.DefaultParams())
-
-			removeInactiveValidatorDelegations(ctx, app)
-			setupICQ(ctx, app)
-			updateMaxSupply(ctx, app)
-
-			addMarkerNavs(ctx, app, GetPioMainnet1DenomToNav())
-
-			setExchangeParams(ctx, app)
-			updateIbcMarkerDenomMetadata(ctx, app)
-
-			return vm, nil
-		},
-		Added: []string{icqtypes.ModuleName, oracletypes.ModuleName, ibchookstypes.StoreKey, hold.ModuleName, exchange.ModuleName},
-	},
-	"tourmaline-rc1": { // upgrade for v1.18.0-rc1
-		Added: []string{ibcratelimit.ModuleName},
-		Handler: func(ctx sdk.Context, app *App, vm module.VersionMap) (module.VersionMap, error) {
-			var err error
-			vm, err = runModuleMigrations(ctx, app, vm)
-			if err != nil {
-				return nil, err
-			}
-
-			removeInactiveValidatorDelegations(ctx, app)
-			convertNavUnits(ctx, app)
-
-			return vm, nil
-		},
-	},
-	"tourmaline-rc2": {}, // upgrade for v1.18.0-rc2
-	"tourmaline-rc3": { // upgrade for v1.18.0-rc3
-		Handler: func(ctx sdk.Context, app *App, vm module.VersionMap) (module.VersionMap, error) {
-			setExchangePaymentParamsToDefaults(ctx, app)
-			addMarkerNavs(ctx, app, GetPioTestnet1DenomToNav())
-			return vm, nil
-		},
-	},
-	"tourmaline": { // upgrade for v1.18.0
-		Added: []string{ibcratelimit.ModuleName},
-		Handler: func(ctx sdk.Context, app *App, vm module.VersionMap) (module.VersionMap, error) {
-			var err error
-			vm, err = runModuleMigrations(ctx, app, vm)
-			if err != nil {
-				return nil, err
-			}
-
-			removeInactiveValidatorDelegations(ctx, app)
-			convertNavUnits(ctx, app)
-			addMarkerNavsWithHeight(ctx, app, GetPioMainnet1NavsTourmaline())
-
-			// This isn't in an rc because it was handled via gov prop for testnet.
-			updateMsgFeesNhashPerMil(ctx, app)
-
-			setExchangePaymentParamsToDefaults(ctx, app)
-
-			return vm, nil
-		},
-	},
```

Comment:
The chain is currently on tourmaline, so those cannot be deleted yet.

Reply:
I'm not sure I understand. We will be upgrading to 1.19 when these changes go in, and we will only be using the 1.19 upgrade handlers during that upgrade. What is preventing us from only having the latest upgrade handler? I've seen a few other chains only support the latest.

Reply:
In looked at Gaia, Evmos, and Wasmd, I’ve noticed they only use the current version into the upgrade process. This makes me curious if there still is a need to include previous versions. However, I’m starting to see potential issues with keeping old versions around. Consider a scenario where someone is manually transitioning the chain through various versions for certain state machine and event updates (we will be doing this at some point). If they’re moving from v1.16 to v1.17 but accidentally use the v1.18 binary, which still has the v1.17 upgrade handler that it runs, it should lead to a halt when trying to achieve consensus, requiring a rollback and potentially a refresh of the data directory from a backup.

If there’s no longer a need to keep these previous versions, the drawbacks seem to outweigh the benefits, including the issue we are facing with increased build times due to my less-than-efficient navigation upgrade handler.

Reply:
It's possible that the SDK fixed this, I haven't checked in a while.

When things are firing up, before it gets to the point where it notices that there's an upgrade, it looks for an upgrade handler for the last upgrade. If it doesn't find it, it complains and stops, preventing code execution from reaching the point where it notices that it's supposed to be upgrading to the next one (that does exist).

This is also the reason why there's a unit test requiring at least two different colors in the upgrade handlers.

I'll run an upgrade test using this branch to see if that's still needed.

Reply:
> I'll run an upgrade test using this branch to see if that's still needed.

Inconclusive. :(

I started with a 1.17.1 chain and used the `tourmaline` upgrade to get it to `1.18.0`. Then used an `umber` upgrade to get it to this branch. When I fired up this branch, I got an error creating the bank keeper, which happens before it can get to any of the upgrade-related stuff.

```shell
> make run
/Users/danielwedul/git/provenance/build/provenanced -t --home /Users/danielwedul/git/provenance/build/run/provenanced start
panic: invalid bank authority address: hrp does not match bech32 prefix: expected 'tp' got 'cosmos': internal logic error [cosmos/cosmos-sdk@v0.50.5/x/auth/codec/bech32_codec.go:37]

goroutine 1 [running]:
github.com/cosmos/cosmos-sdk/x/bank/keeper.NewBaseKeeper({_, _}, {_, _}, {_, _}, _, {_, _}, {_, ...})
	github.com/cosmos/cosmos-sdk@v0.50.5/x/bank/keeper/keeper.go:94 +0x214
github.com/provenance-io/provenance/app.New({0x10653a3a8?, 0x140018a44d0}, {0x10654f948, 0x140012d4040}, {0x0, 0x0}, 0x1, 0x140017d0570?, {0x16d17905e, 0x37}, ...)
	github.com/provenance-io/provenance/app/app.go:433 +0x1040
github.com/provenance-io/provenance/cmd/provenanced/cmd.newApp({0x10653a3a8, 0x140018a44d0}, {0x10654f948, 0x140012d4040}, {0x0, 0x0}, {0x1064f4200?, 0x140012ca1c0?})
	github.com/provenance-io/provenance/cmd/provenanced/cmd/root.go:370 +0x694
github.com/cosmos/cosmos-sdk/server.startApp(0x140019757a0, 0x1064e0968, {0x1064e1188?, 0x0?, 0x0?})
	github.com/cosmos/cosmos-sdk@v0.50.5/server/start.go:614 +0x1bc
github.com/cosmos/cosmos-sdk/server.start(_, {{0x0, 0x0, 0x0}, {0x1065551b0, 0x14001bc3920}, 0x0, {0x14001c01790, 0x7}, {0x106571fe0, ...}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.50.5/server/start.go:216 +0xd8
github.com/cosmos/cosmos-sdk/server.StartCmdWithOptions.func2.1()
	github.com/cosmos/cosmos-sdk@v0.50.5/server/start.go:190 +0x58
github.com/cosmos/cosmos-sdk/server.wrapCPUProfile(0x140019757a0, 0x14001c9f808)
	github.com/cosmos/cosmos-sdk@v0.50.5/server/start.go:566 +0x16c
github.com/cosmos/cosmos-sdk/server.StartCmdWithOptions.func2(0x140010ba800?, {0x14001b65830?, 0x0?, 0x3?})
	github.com/cosmos/cosmos-sdk@v0.50.5/server/start.go:189 +0x194
github.com/spf13/cobra.(*Command).execute(0x1400186f200, {0x14001b657d0, 0x3, 0x3})
	github.com/spf13/cobra@v1.8.0/command.go:983 +0x82c
github.com/spf13/cobra.(*Command).ExecuteC(0x1400182c000)
	github.com/spf13/cobra@v1.8.0/command.go:1115 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	github.com/spf13/cobra@v1.8.0/command.go:1032
github.com/provenance-io/provenance/cmd/provenanced/cmd.Execute(0x140000b6701?)
	github.com/provenance-io/provenance/cmd/provenanced/cmd/root.go:188 +0x1fc
main.main()
	github.com/provenance-io/provenance/cmd/provenanced/main.go:13 +0x24
make: *** [run] Error 2
```

---

As far as I know, the content of the old upgrade handlers didn't matter, only that there was an entry for it. Since `mainnet` last did `tourmaline` and `testnet` last did `tourmaline-rc3`, let's put empty entries for those two back into this map. We still get to delete the rest of the stuff, but if an entry is still required for the current entry, we'll still have it. Once we get more of the kinks worked out with v0.50, we can try removing those empty entries again to see if they're actually still needed.

Reply:
Main isn't stable so we unable to test it there. I tested it with v1.17.0 and I was able to upgrade successfully without the current upgrade handler defined.

```
var upgrades = map[string]appUpgrade{
	"next2": { // upgrade for next
		Handler: func(ctx sdk.Context, app *App, vm module.VersionMap) (module.VersionMap, error) {
			var err error
			vm, err = runModuleMigrations(ctx, app, vm)
			if err != nil {
				return nil, err
			}

			return vm, nil
		},
	},
	// TODO - Add new upgrade definitions here.
}
```

```
2:38PM INF Managing upgrade lastHeight=774 plan=next2 upgradeHeight=775
2:38PM INF No store upgrades required height=775 plan=next2
2:38PM INF starting node with ABCI Tendermint in-process
2:38PM INF service start impl=multiAppConn module=proxy msg={}
2:38PM INF service start connection=query impl=localClient module=abci-client msg={}
2:38PM INF service start connection=snapshot impl=localClient module=abci-client msg={}
2:38PM INF service start connection=mempool impl=localClient module=abci-client msg={}
2:38PM INF service start connection=consensus impl=localClient module=abci-client msg={}
2:38PM INF service start impl=EventBus module=events msg={}
2:38PM INF service start impl=PubSub module=pubsub msg={}
2:38PM INF service start impl=IndexerService module=txindex msg={}
2:38PM INF ABCI Handshake App Info hash="\x7f� I�\x1b�\f1���p\x1b'\x16�)ia�V�eG2���h��" height=774 module=consensus protocol-version=0 software-version=v1.17.0
2:38PM INF ABCI Replay Blocks appHeight=774 module=consensus stateHeight=774 storeHeight=775
2:38PM INF Replay last block using real app module=consensus
2:38PM INF applying upgrade "next2" at height: 775
2:38PM INF Starting upgrade to "next2" version-map={"attribute":2,"auth":3,"authz":2,"bank":4,"capability":1,"crisis":1,"distribution":2,"evidence":1,"exchange":1,"feegrant":2,"genutil":1,"gov":3,"group":1,"hold":1,"ibc":2,"ibchooks":1,"interchainaccounts":2,"interchainquery":1,"marker":2,"metadata":3,"mint":1,"msgfees":1,"name":2,"oracle":1,"params":1,"quarantine":1,"reward":1,"sanction":1,"slashing":2,"staking":3,"transfer":2,"trigger":1,"upgrade":2,"vesting":1,"wasm":2}
2:38PM INF Starting module migrations. This may take a significant amount of time to complete. Do not restart node.
2:38PM INF Module migrations completed.
2:38PM INF Successfully upgraded to "next2" version-map={"attribute":2,"auth":3,"authz":2,"bank":4,"capability":1,"crisis":1,"distribution":2,"evidence":1,"exchange":1,"feegrant":2,"genutil":1,"gov":3,"group":1,"hold":1,"ibc":2,"ibchooks":1,"interchainaccounts":2,"interchainquery":1,"marker":2,"metadata":3,"mint":1,"msgfees":1,"name":2,"oracle":1,"params":1,"quarantine":1,"reward":1,"sanction":1,"slashing":2,"staking":3,"transfer":2,"trigger":1,"upgrade":2,"vesting":1,"wasm":2}
2:38PM INF executed block height=775 module=consensus num_invalid_txs=0 num_valid_txs=0
2:38PM INF commit synced commit=436F6D6D697449447B5B313737203130203135392031313020313037203232322032333020323332203920313734203239203131203130362032323920323430203630203231382031373920323239203130372031343720353420363420353420313635203138362031313420363020313931203931203134372039375D3A3330377D
```

I'm still convinced that we should remove the unused handlers especially after that test.

Reply:
@SpicyLemon I believe that is a main branch error.  I was seeing that when running some of the tests after getting things wired in for v50.

Reply:
Sounds good to me. I'm sure we'll do some upgrade testing prior to any release. So if they are still required, we'll find out then and can re-add some empty entries if needed.